units/commit_queue.go

```diff
@@ -104,165 +104,175 @@ func (j *commitQueueJob) Run(ctx context.Context) {
-			"job":     commitQueueJobName,
-			"source":  "commit queue",
-			"project": j.QueueID,
-			"item":    nextItem,
-			"message": "can't dequeue item",
-		}))
-
+	if nextItem == nil {
 		return
 	}
+	j.AddError(errors.Wrap(cq.SetProcessing(true), "can't set processing to true"))
 
 	conf := j.env.Settings()
 	githubToken, err := conf.GetGithubOauthToken()
 	if err != nil {
 		j.AddError(errors.Wrap(err, "can't get github token"))
+		j.AddError(errors.Wrap(cq.SetProcessing(false), "can't set processing to false"))
 		return
 	}
 
-	pr, err := thirdparty.GetGithubPullRequest(ctx, githubToken, projectRef.Owner, projectRef.Repo, nextItemInt)
+	pr, dequeue, err := checkPR(ctx, githubToken, nextItem.Issue, projectRef.Owner, projectRef.Repo)
 	if err != nil {
-		j.AddError(errors.Wrap(err, "can't get PR from GitHub"))
-		grip.Error(message.WrapError(err, message.Fields{
-			"job":     commitQueueJobName,
-			"source":  "commit queue",
-			"project": j.QueueID,
-			"item":    nextItem,
-			"message": "can't get PR from GitHub",
-		}))
+		j.logError(err, "PR not valid for merge", nextItem)
+		if dequeue {
+			if pr != nil {
+				j.AddError(sendCommitQueueGithubStatus(j.env, pr, message.GithubStateFailure, "PR not valid for merge", ""))
+			}
+			j.dequeue(cq, nextItem)
+		} else {
+			j.AddError(errors.Wrap(cq.SetProcessing(false), "can't set processing to false"))
+		}
 		return
 	}
 
-	if err = validatePR(pr); err != nil {
-		j.AddError(errors.Wrap(err, "invalid PR"))
-		grip.Error(message.WrapError(err, message.Fields{
-			"job":     commitQueueJobName,
-			"source":  "commit queue",
-			"project": j.QueueID,
-			"item":    nextItem,
-			"message": "invalid PR",
-		}))
-
-		_, err = cq.Remove(nextItem)
-		j.AddError(errors.Wrapf(err, "error dequeuing item '%s'", nextItem))
-		grip.Error(message.WrapError(err, message.Fields{
-			"job":     commitQueueJobName,
-			"source":  "commit queue",
-			"project": j.QueueID,
-			"item":    nextItem,
-			"message": "can't dequeue item",
-		}))
+	patchDoc, projectConfig, err := makeMergePatch(ctx, pr, githubToken, projectRef.Identifier)
+	if err != nil {
+		j.logError(err, "can't make patch", nextItem)
+		j.AddError(sendCommitQueueGithubStatus(j.env, pr, message.GithubStateFailure, "can't make patch", ""))
+		j.dequeue(cq, nextItem)
+		j.AddError(errors.Wrap(cq.SetProcessing(false), "can't set processing to false"))
 		return
 	}
 
-	// Remove already merged PRs from the queue
-	if *pr.Merged {
-		_, err = cq.Remove(nextItem)
-		j.AddError(errors.Wrapf(err, "error dequeuing item '%s'", nextItem))
-		grip.Error(message.WrapError(err, message.Fields{
-			"job":     commitQueueJobName,
-			"source":  "commit queue",
-			"project": j.QueueID,
-			"item":    nextItem,
-			"message": "can't dequeue item",
-		}))
+	modulePRs, modulePatches, dequeue, err := getModules(ctx, githubToken, nextItem, projectConfig)
+	if err != nil {
+		j.logError(err, "can't get modules", nextItem)
+		if dequeue {
```

Comment:
what if dequeue is true and error is not nil, you'll miss an error?

Reply:
getModules will only ask for dequeue if there's an error.
We only want to dequeue for some types of errors, and dequeue tells us which.